// coinbase-pro-node/dist/account/AccountAPI.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
class AccountAPI {
    constructor(apiClient) {
        this.apiClient = apiClient;
    }
    getAccount(accountId) {
        return __awaiter(this, void 0, void 0, function* () {
            const resource = `${AccountAPI.URL.ACCOUNTS}/${accountId}`;
            const response = yield this.apiClient.get(resource);
            return response.data;
        });
    }
    getAccountHistory(accountId) {
        return __awaiter(this, void 0, void 0, function* () {
            const resource = `${AccountAPI.URL.ACCOUNTS}/${accountId}/ledger`;
            const response = yield this.apiClient.get(resource);
            return response.data;
        });
    }
    getHolds(accountId) {
        return __awaiter(this, void 0, void 0, function* () {
            const resource = `${AccountAPI.URL.ACCOUNTS}/${accountId}/holds`;
            const response = yield this.apiClient.get(resource);
            return response.data;
        });
    }
    listAccounts() {
        return __awaiter(this, void 0, void 0, function* () {
            const resource = AccountAPI.URL.ACCOUNTS;
            const response = yield this.apiClient.get(resource);
            return response.data;
        });
    }
}
exports.AccountAPI = AccountAPI;
AccountAPI.URL = {
    ACCOUNTS: `/accounts`,
};
