// coinbase-pro-node/dist/profile/ProfileAPI.test.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const nock_1 = __importDefault(require("nock"));
const ProfileAPI_1 = require("./ProfileAPI");
describe('ProfileAPI', () => {
    afterEach(() => nock_1.default.cleanAll());
    describe('listProfiles', () => {
        it('lists your profiles', () => __awaiter(void 0, void 0, void 0, function* () {
            const profileId = '86602c68-306a-4500-ac73-4ce56a91d83c';
            nock_1.default(global.REST_URL)
                .get(ProfileAPI_1.ProfileAPI.URL.PROFILES)
                .query(() => true)
                .reply(200, JSON.stringify([
                {
                    active: true,
                    created_at: '2019-11-18T15:08:40.236309Z',
                    id: profileId,
                    is_default: true,
                    name: 'default',
                    user_id: '5844eceecf7e803e259d0365',
                },
            ]));
            const profiles = yield global.client.rest.profile.listProfiles();
            expect(profiles.length).toBe(1);
            expect(profiles[0].id).toBe(profileId);
        }));
    });
    describe('getProfile', () => {
        it('returns a single profile by profile ID', () => __awaiter(void 0, void 0, void 0, function* () {
            const profileId = '86602c68-306a-4500-ac73-4ce56a91d83c';
            nock_1.default(global.REST_URL)
                .get(`${ProfileAPI_1.ProfileAPI.URL.PROFILES}/${profileId}`)
                .query(() => true)
                .reply(200, JSON.stringify({
                active: true,
                created_at: '2019-11-18T15:08:40.236309Z',
                id: profileId,
                is_default: true,
                name: 'default',
                user_id: '5844eceecf7e803e259d0365',
            }));
            const profile = yield global.client.rest.profile.getProfile(profileId);
            expect(profile.id).toBe(profileId);
        }));
        it('returns null if a profile cannot be found', () => __awaiter(void 0, void 0, void 0, function* () {
            const profileId = '86602c68-306a-4500-ac73-4ce56a91d83d';
            nock_1.default(global.REST_URL)
                .get(`${ProfileAPI_1.ProfileAPI.URL.PROFILES}/${profileId}`)
                .query(() => true)
                .reply(404);
            const profile = yield global.client.rest.profile.getProfile(profileId);
            expect(profile).toEqual(null);
        }));
        it('rethrows errors with status code other than 404', () => __awaiter(void 0, void 0, void 0, function* () {
            const profileId = '86602c68-306a-4500-ac73-4ce56a91d83d';
            nock_1.default(global.REST_URL)
                .get(`${ProfileAPI_1.ProfileAPI.URL.PROFILES}/${profileId}`)
                .query(() => true)
                .reply(500);
            try {
                yield global.client.rest.profile.getProfile(profileId);
            }
            catch (error) {
                expect(error.response.status).toBe(500);
            }
        }));
    });
    describe('transferFunds', () => {
        it('transfers funds from one profile to another', (done) => __awaiter(void 0, void 0, void 0, function* () {
            nock_1.default(global.REST_URL)
                .post(`${ProfileAPI_1.ProfileAPI.URL.PROFILES}/transfer`)
                .query(() => true)
                .reply(200);
            const transfer = {
                amount: '1000.00',
                currency: 'BTC',
                from: '86602c68-306a-4500-ac73-4ce56a91d83c',
                to: 'e87429d3-f0a7-4f28-8dff-8dd93d383de1',
            };
            yield global.client.rest.profile.transferFunds(transfer);
            done();
        }));
    });
});
