// coinbase-pro-node/dist/client/RESTClient.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const axios_1 = __importDefault(require("axios"));
const axios_rate_limit_1 = __importDefault(require("axios-rate-limit"));
const AccountAPI_1 = require("../account/AccountAPI");
const RequestSigner_1 = require("../auth/RequestSigner");
const OrderAPI_1 = require("../order/OrderAPI");
const ProductAPI_1 = require("../product/ProductAPI");
const TimeAPI_1 = require("../time/TimeAPI");
const UserAPI_1 = require("../user/UserAPI");
const FillAPI_1 = require("../fill/FillAPI");
const querystring_1 = __importDefault(require("querystring"));
const ProfileAPI_1 = require("../profile/ProfileAPI");
class RESTClient {
    constructor(baseURL, auth) {
        this.httpClient = axios_rate_limit_1.default(axios_1.default.create({
            baseURL: baseURL,
            timeout: 5000,
        }), { maxRequests: 3, perMilliseconds: 1000 });
        this.httpClient.interceptors.request.use((config) => __awaiter(this, void 0, void 0, function* () {
            const baseURL = String(config.baseURL);
            const url = String(config.url);
            const requestPath = url.replace(baseURL, '');
            const clockSkew = yield TimeAPI_1.TimeAPI.getClockSkew(baseURL);
            const signedRequest = RequestSigner_1.RequestSigner.signRequest(auth, {
                httpMethod: String(config.method).toUpperCase(),
                payload: RESTClient.stringifyPayload(config),
                requestPath,
            }, clockSkew);
            config.headers = Object.assign(Object.assign({}, config.headers), { 'CB-ACCESS-KEY': signedRequest.key, 'CB-ACCESS-PASSPHRASE': signedRequest.passphrase, 'CB-ACCESS-SIGN': signedRequest.signature, 'CB-ACCESS-TIMESTAMP': signedRequest.timestamp });
            return config;
        }));
        this.account = new AccountAPI_1.AccountAPI(this.httpClient);
        this.fill = new FillAPI_1.FillAPI(this.httpClient);
        this.order = new OrderAPI_1.OrderAPI(this.httpClient);
        this.product = new ProductAPI_1.ProductAPI(this.httpClient);
        this.profile = new ProfileAPI_1.ProfileAPI(this.httpClient);
        this.user = new UserAPI_1.UserAPI(this.httpClient);
    }
    get defaults() {
        return this.httpClient.defaults;
    }
    get interceptors() {
        return this.httpClient.interceptors;
    }
    static stringifyPayload(config) {
        if (config.data) {
            return JSON.stringify(config.data);
        }
        const params = querystring_1.default.stringify(config.params);
        return params ? `?${params}` : '';
    }
}
exports.RESTClient = RESTClient;
