// coinbase-pro-node/dist/client/RESTClient.test.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const nock_1 = __importDefault(require("nock"));
const AccountAPI_1 = require("../account/AccountAPI");
const GET_200_json_1 = __importDefault(require("../test/fixtures/rest/accounts/GET-200.json"));
const RESTClient_1 = require("./RESTClient");
describe('RESTClient', () => {
    function createClient() {
        const auth = {
            apiKey: '',
            apiSecret: '',
            passphrase: '',
            useSandbox: true,
        };
        const baseURL = global.REST_URL;
        return new RESTClient_1.RESTClient(baseURL, auth);
    }
    describe('defaults', () => {
        it('supports overriding the timeout limit', () => {
            const client = createClient();
            expect(client.defaults.timeout).toBe(5000);
            client.defaults.timeout = 2500;
            expect(client.defaults.timeout).toBe(2500);
        });
    });
    describe('interceptors', () => {
        afterAll(() => nock_1.default.cleanAll());
        beforeAll(() => {
            nock_1.default(global.REST_URL)
                .get(AccountAPI_1.AccountAPI.URL.ACCOUNTS)
                .query(() => true)
                .reply(() => [200, JSON.stringify(GET_200_json_1.default)])
                .persist();
        });
        it('supports custom HTTP interceptors', () => __awaiter(void 0, void 0, void 0, function* () {
            const client = createClient();
            const onRequest = jasmine.createSpy('onRequest').and.callFake((config) => {
                return config;
            });
            const myInterceptor = client.interceptors.request.use(onRequest);
            yield client.account.listAccounts();
            expect(onRequest).toHaveBeenCalledTimes(1);
            client.interceptors.request.eject(myInterceptor);
            yield client.account.listAccounts();
            expect(onRequest).toHaveBeenCalledTimes(1);
        }));
    });
});
